### PR TITLE
bpo-40059: tomllib

### PR DESCRIPTION
**Pull request by hukkin**
This adds a new standard library module, `tomllib`, for parsing TOML. The recently accepted [PEP 680 -- tomllib](https://www.python.org/dev/peps/pep-0680/) is relevant here.

This PR has already seen some review in a PR under my personal fork: https://github.com/hukkin/cpython/pull/2 (thanks to @encukou, @merwok, @hauntsaninja, @JelleZijlstra (I hope I'm not forgetting anyone)).

The implementation is based on [Tomli](https://github.com/hukkin/tomli) which I plan to keep maintaining as a backport for Python versions 3.7, 3.8, 3.9 and 3.10, until finally Python 3.10 goes EOL.

## Steps taken (converting `tomli` to `tomllib`)
- Move everything in `tomli:src/tomli` to `Lib/tomllib`. Exclude `py.typed`.
- Remove `__version__ = ...` line from `Lib/tomllib/__init__.py`
- Move everything in `tomli:tests` to `Lib/test/test_tomllib`. Exclude the following test data dirs recursively:
  - `tomli:tests/data/invalid/_external/`
  - `tomli:tests/data/valid/_external/`
- Create `Lib/test/test_tomllib/__main__.py`:

  ```python
  import unittest

  from . import load_tests


  unittest.main()
  ```

- Add the following to `Lib/test/test_tomllib/__init__.py`:

  ```python
  import os
  from test.support import load_package_tests

  def load_tests(*args):
      return load_package_tests(os.path.dirname(__file__), *args)
  ```
  
  Also change `import tomli as tomllib` to `import tomllib`.

- In `cpython/Lib/tomllib/_parser.py` replace `__fp` with `fp` and `__s` with `s`. Add the `/` to `load` and `loads` function signatures.

- Run `make regen-stdlib-module-names`

- Create `Doc/library/tomllib.rst` and reference it in `Doc/library/fileformats.rst`

edit: For reference, there's one more step – [Add `tomllib` to Makefile.pre.in](https://github.com/python/cpython/pull/31784/files)

https://bugs.python.org/issue40059

### CODE REVIEW

**Comment by hukkin** [comment]
A question: upstream ([Tomli](https://github.com/hukkin/tomli)) is formatted with Black, using Black's defaults. This means a line length of 88. Should I reformat with line length at 79?

**Comment by mgorny** [comment]
Perhaps it'd make sense to include the tomli version used (or even the git commit hash), to make future syncing easier.

**Comment by hukkin** [comment]
> Perhaps it'd make sense to include the tomli version used (or even the git commit hash), to make future syncing easier.

It was previously suggested that I remove Tomli version from the standard library port: https://github.com/hukkin/cpython/pull/2#issuecomment-1023366439

I plan to include a migration guide in Tomli repository, also including commit hash that tracks what is currently in the stdlib.

**Comment by TeamSpen210** [comment]
Would it be good to mention in the docs why `load()` takes only binary files? The encoding requirement probably isn't obvious for first-time users.

**Comment by encukou** [comment]
I should get to the review this week or the next.

> A question: upstream ([Tomli](https://github.com/hukkin/tomli)) is formatted with Black, using Black's defaults. This means a line length of 88. Should I reformat with line length at 79?

No, I don't think that's worth it. And anyway, line length is not the only point where Black disagrees with PEP8 (starting with, like, the [whole philosophy](https://www.python.org/dev/peps/pep-0008/#a-foolish-consistency-is-the-hobgoblin-of-little-minds)). 
But future edits probably won't use Black.

> Would it be good to mention in the docs why load() takes only binary files? The encoding requirement probably isn't obvious for first-time users.

Let's leave docs improvements to future PRs, so they get a proper discussion and don't delay this PR?

File: .github/CODEOWNERS
Comment (encukou) [inline .github/CODEOWNERS]: This file can only list core devs, sadly.
```suggestion
**/*tomllib*                  @encukou
```

File: Lib/test/test_tomllib/test_data.py
Comment (encukou) [inline Lib/test/test_tomllib/test_data.py]: `MissingFile` looks unnecessary. Does tomli need it?

File: Lib/test/test_tomllib/test_data.py
Comment (encukou) [inline Lib/test/test_tomllib/test_data.py:36]: For peace of mind, could you assert `len(VALID_FILES) > 0`, and same for `INVALID_FILES`?

File: Lib/test/test_tomllib/test_data.py
Comment (encukou) [reply Lib/test/test_tomllib/test_data.py]: Ah, I see it does. 
For a poor man's xfail, you could you assert that `p.stem` is one of the expected failing cases.

**Review by encukou — APPROVED** [review]
I have a few test nitpicks, but I'm happy to merge this!

File: Lib/test/test_tomllib/test_data.py
Comment (hukkin) [reply Lib/test/test_tomllib/test_data.py:36]: :+1: Makes sense. Added the asserts.

File: Lib/test/test_tomllib/test_data.py
Comment (hukkin) [reply Lib/test/test_tomllib/test_data.py]: > MissingFile looks unnecessary. Does tomli need it?

Yeah one of the two external test suites has a couple test cases where the expected data is missing.

> For a poor man's xfail, you could you assert that p.stem is one of the expected failing cases.

I'm not sure I understand what you mean. Maybe you can show with the "Add a suggestion" feature?

Comment (hauntsaninja) [reply Lib/test/test_tomllib/test_data.py]: I think something like:
```
if isinstance(expected, MissingFile):
    assert valid.stem in ("xfail_test_case1, "xfail_test_case2", ...)
    continue
```

Comment (hukkin) [reply Lib/test/test_tomllib/test_data.py]: Oh yes, of course, thanks. Committed that.

**Comment by hukkin** [comment]
It seems the failing CI job (`AMD64 Arch Linux Usan PR`) errors in [other PRs too](https://github.com/python/cpython/pull/31640#issuecomment-1058275118) so should be unrelated to this PR.

**Comment by encukou** [comment]
The alpha 6 release is a bit bumpy and I don't want to destabilize it, so I'm holding off the merge until it's out.
I tested with *all* buildbots to see if there's an unforeseen platform-specific issue. It's common to see a few buildbots fail.

**Comment by encukou** [comment]
Let's get it in!

**Pull request merged**
